In [1]:
import pandas as pd

In [2]:
tariff_df = pd.read_csv('Tariff_20240104.csv', encoding='cp949', dtype=str)  # " 공공데이터 포털 기준 최종 수정일 2024-01-04"

In [163]:
# (tariff_df['품목번호'].str.len() <= 9).sum()

0

### tariff에서 필요한 feature만 select

In [3]:
new_df = tariff_df[['품목번호','관세율','관세율구분','용도세율구분','적용개시일','적용만료일']]

new_df.head()

품목번호 관세율 관세율구분 용도세율구분       적용개시일       적용만료일
0  0101211000   0     A    NaN  2024-01-01  2024-12-31
1  0101219000   8     A    NaN  2024-01-01  2024-12-31
2  0101291000   8     A    NaN  2024-01-01  2024-12-31
3  0101299000   8     A    NaN  2024-01-01  2024-12-31
4  0101300000   8     A    NaN  2024-01-01  2024-12-31

In [4]:
hscode_df = pd.read_csv('HS코드(2024).csv', dtype=str, encoding='cp949')

### hscode_df에 '2424000000' 존재

### hscode_df 내 'HS부호' 9자리인 data 다수 확인 됨. 
#### 실제로 9자리 이하의 HS부호가 사용된다면, 현재 10자리만을 고려하는 로직과 충돌할 수 있음

In [22]:
# import requests
# import pandas as pd
# from io import StringIO
# import json

# # 요청 매개변수 설정
# params = {
#     'serviceKey': 'gGDom+BIbgsPl1jf3j7m6aMaMtBwgPBtgrenGKVs8VX2Ap9RcGYULywUHIvwxp2bI/XJ9Z7DJMqwuO8RRVMofw==',
#     'page': '1',
#     'perPage': '1000'  # 한 번의 요청으로 얻을 수 있는 최대 항목 수
# }

# # 요청 URL 설정
# url = 'https://api.odcloud.kr/api/15049722/v1/uddi:1914a5f2-6db7-4b78-9495-cd092b6e4029?page=1&perPage=1000'

# # 첫 번째 요청을 보내서 총 항목 수를 확인
# response = requests.get(url, params=params)
# data = json.loads(response.text)

In [23]:
# data

{'currentCount': 1000,
 'data': [{'HS부호': 101211000,
   '규격사항내용': None,
   '규격설명': None,
   '성질통합분류코드': 11020101,
   '성질통합분류코드명': '(말)',
   '수량단위최대단가': None,
   '수량단위코드': 'U',
   '수입성질코드': '12399',
   '수출성질코드': '11Z01',
   '영문품목명': 'For farm breeding',
   '적용시작일자': '2013-01-01',
   '적용종료일자': '2022-12-31',
   '중량단위최대단가': None,
   '중량단위코드': 'KG',
   '참고규격명': None,
   '품목규격명': None,
   '필수규격명': None,
   '한국표준무역분류명': None,
   '한글품목명': '농가 사육용'},
  {'HS부호': 101219000,
   '규격사항내용': None,
   '규격설명': None,
   '성질통합분류코드': 11020101,
   '성질통합분류코드명': '(말)',
   '수량단위최대단가': None,
   '수량단위코드': 'U',
   '수입성질코드': '12399',
   '수출성질코드': '11Z01',
   '영문품목명': 'Other',
   '적용시작일자': '2012-01-01',
   '적용종료일자': '2022-12-31',
   '중량단위최대단가': None,
   '중량단위코드': 'KG',
   '참고규격명': None,
   '품목규격명': None,
   '필수규격명': None,
   '한국표준무역분류명': None,
   '한글품목명': '기타'},
  {'HS부호': 101291000,
   '규격사항내용': None,
   '규격설명': None,
   '성질통합분류코드': 11020101,
   '성질통합분류코드명': '(말)',
   '수량단위최대단가': None,
   '수량단위코드': 'U',
   '수입성질코드

In [ ]:
# total_count = data['totalCount']

# # 총 페이지 수 계산
# total_pages = total_count // int(params['perPage']) + (total_count % int(params['perPage']) > 0)

# # 모든 페이지에 대해 요청을 반복
# all_data = []
# for page in range(1, total_pages + 1):
#     params['page'] = str(page)
#     response = requests.get(url, params=params)
#     df_page = pd.read_json(StringIO(response.text), dtype=str)
#     all_data.append(df_page)

# # all_data에는 이제 API가 제공하는 모든 데이터가 포함되어 있습니다.
# # 이 데이터를 하나의 DataFrame으로 합칩니다.
# df = pd.concat(all_data, ignore_index=True)

In [7]:
tariff_df.shape

(367969, 9)

### hscode_df의 feature 파악

In [8]:
hscode_df.columns 

Index(['HS부호', '적용시작일자', '적용종료일자', '한글품목명', '영문품목명', 'HS부호내용', '한국표준무역분류명',
       '수량단위최대단가', '중량단위최대단가', '수량단위코드', '중량단위코드', '수출성질코드', '수입성질코드', '품목규격명',
       '필수규격명', '참고규격명', '규격설명', '규격사항내용', '성질통합분류코드', '성질통합분류코드명'],
      dtype='object')

### hscode_df 필요 feature select

In [9]:
hscode_df = hscode_df[['HS부호', '적용시작일자', '적용종료일자', '한글품목명', '영문품목명', '수량단위코드', '중량단위코드', '수출성질코드', '수입성질코드', '성질통합분류코드', '성질통합분류코드명']]

hscode_df.head()

HS부호    적용시작일자    적용종료일자   한글품목명              영문품목명 수량단위코드 중량단위코드  \
0  0101211000  20130101  20241231  농가 사육용  For farm breeding      U     KG   
1  0101219000  20120101  20241231      기타              Other      U     KG   
2  0101291000  20120101  20241231     경주말  Horses for racing      U     KG   
3  0101299000  20120101  20241231      기타              Other      U     KG   
4  0101300000  20220101  20241231     당나귀              Asses      U     KG   

  수출성질코드 수입성질코드  성질통합분류코드  성질통합분류코드명  
0  11Z01  12399  11020101        (말)  
1  11Z01  12399  11020101        (말)  
2  11Z01  12399  11020101        (말)  
3  11Z01  12399  11020101        (말)  
4  11Z01  12399  11020190  (기타 산 동물)

In [10]:
hscode_df = hscode_df.rename(columns = {'HS부호': '품목번호'})

### 우선 HSCODE표에는 6자리 이하의 코드는 존재하지 않는다.
### 7자리 이상부터 존재하는데 실무에서 7자리만으로도 사용되는 지, 혹은 7자리인경우 뒤에 000을 붙여 사용하는지 확인이 필요.(혹은 기타로 취급할 수도 있음)

In [19]:
#print(hscode_df[hscode_df['품목번호']== '2424000000'])
print(hscode_df[hscode_df['품목번호'].str.len() <= 7])
#print(hscode_df['품목번호'].str.len() <= 7).sum()

          품목번호    적용시작일자    적용종료일자  \
160    0207603  20160101  20241231   
165    0207604  20160101  20241231   
648    0307992  20160101  20241231   
652    0307994  20170101  20241231   
716    0404102  20160101  20241231   
791    0507901  20160101  20241231   
1363   1211201  20170101  20241231   
1382   1211202  20170101  20241231   
1391   1211203  20170101  20241231   
1394   1211209  20160101  20241231   
1405   1211901  20170101  20241231   
1559   1302191  20160101  20241231   
1905   1806902  20160101  20241231   
2627   2710197  20160101  20241231   
2669   2710209  20160101  20241231   
3535   2920909  20220101  20241231   
3690   2927001  20160101  20241231   
3693   2927002  20160101  20241231   
3912   2937902  20160101  20241231   
4030   3004209  20160101  20241231   
4063   3004909  20160101  20241231   
4302   3301904  20160101  20241231   
4417   3501901  20160101  20241231   
4494   3701919  20160101  20241231   
4499   3701999  20160101  20241231   
4545   37040

In [20]:
tariff_df[tariff_df['품목번호']== '2424000000']

Empty DataFrame
Columns: [품목번호, 관세율구분, 관세율, 단위당세액, 기준가격, 적용국가구분, 용도세율구분, 적용개시일, 적용만료일]
Index: []

In [21]:
hscode_df[hscode_df['품목번호']== '2424000000']

품목번호    적용시작일자    적용종료일자 한글품목명            영문품목명 수량단위코드 중량단위코드  \
2355  2424000000  20020101  20241231  이사화물  household goods    NaN     KG   

     수출성질코드 수입성질코드  성질통합분류코드  성질통합분류코드명  
2355  11Z01  14Z01  14010101  (간이세율적용분)

In [23]:
len(tariff_df.merge(hscode_df, on = '품목번호', how = 'left').columns)

19

In [24]:
len(hscode_df.columns) + len(tariff_df.columns) == len(tariff_df.merge(hscode_df, on = '품목번호', how = 'left').columns) +1

True

In [25]:
len(hscode_df)

12422

In [26]:
# mergedHsTariff = tariff_df.merge(hscode_df, on = '품목번호', how = 'left')
mergedHsTariff = hscode_df.merge(tariff_df, on = '품목번호', how = 'left')

In [27]:
mergedHsTariff = mergedHsTariff.drop(columns = ['기준가격'])

In [28]:
len(mergedHsTariff.columns)

18

In [29]:
cols = ['품목번호', '한글품목명', '영문품목명','관세율','관세율구분', '성질통합분류코드명'] + [col for col in mergedHsTariff.columns if col not in ['품목번호', '한글품목명', '영문품목명','관세율' ,'관세율구분','성질통합분류코드명']]
mergedHsTariff = mergedHsTariff[cols]

In [193]:
mergedHsTariff[mergedHsTariff['품목번호']== '0101200000']

Empty DataFrame
Columns: [품목번호, 한글품목명, 영문품목명, 관세율, 관세율구분, 성질통합분류코드명, 적용시작일자, 적용종료일자, 수량단위코드, 중량단위코드, 수출성질코드, 수입성질코드, 성질통합분류코드, 단위당세액, 적용국가구분, 용도세율구분, 적용개시일, 적용만료일]
Index: []

In [32]:
#len(mergedHsTariff[mergedHsTariff['품목번호'].str.len() == 10])
len(mergedHsTariff[mergedHsTariff['품목번호'].str.len() == 9])

170

In [33]:
len(mergedHsTariff[mergedHsTariff['품목번호'].str.len() == 10])

367970

In [34]:
len(mergedHsTariff) == len(mergedHsTariff[mergedHsTariff['품목번호'].str.len() == 10]) + len(mergedHsTariff[mergedHsTariff['품목번호'].str.len() <= 9])

True

In [88]:
unique_count_hscode_under9 = hscode_df[hscode_df['품목번호'].str.len() <= 9]['품목번호'].nunique()
unique_count_hscode = hscode_df['품목번호'].nunique()
unique_count_tariff = tariff_df['품목번호'].nunique()
unique_count_merged = mergedHsTariff['품목번호'].nunique()
print("HSK 22년 기준 11293개(12,242개에서 341개 신설, 1290개 삭제),  ",unique_count_merged)

HSK 22년 기준 11293개(12,242개에서 341개 신설, 1290개 삭제),   12422


HSK는 세계관세기구가 제정한 6단위 코드로 구성된 국제 통일 상품분류체계(HS)를 바탕으로 우리나라의 필요에 따라 품목을 세분화한 10단위 품목분류표인데, 이번 개정으로 HSK상 품목이 12,242개에서 11,293개(신설 341개, 삭제 1,290개)로 변경되었습니다.

In [97]:
print("unique_count_hscode_under9: ", unique_count_hscode_under9)
print("unique_count_hscode: ",unique_count_hscode)
print("unique_count_tariff: ", unique_count_tariff, ", tariff에는 9자리 이하의 hsk는 존재하지 않음")
print("hscod에서 tariff를 뺀 값: ",unique_count_hscode-unique_count_tariff)
print("1개는 '2424000000', 결론 : 22년 hscode 기준인 11293과 일치함 ")

print(" ")
print("len(hscode_df)- unique_count_hscode :",len(hscode_df)- unique_count_hscode)
print("len(hscode_df)- unique_count_tariff :",len(hscode_df)- unique_count_tariff) # 1129개가 어디서 나왔는가?
print("len(hscode_df)- unique_count_merged :",len(hscode_df)- unique_count_merged)
print("len(mergedHsTariff) - len(tariff_df) :",len(mergedHsTariff) - len(tariff_df)) # HSCODE 기준으로 left join했기 때문에 1129개가 나옴
print("") 

unique_count_hscode_under9:  1128
unique_count_hscode:  12422
unique_count_tariff:  11293 , tariff에는 9자리 이하의 hsk는 존재하지 않음
hscod에서 tariff를 뺀 값:  1129
1개는 '2424000000', 결론 : 22년 hscode 기준인 11293과 일치함 
 
len(hscode_df)- unique_count_hscode : 0
len(hscode_df)- unique_count_tariff : 1129
len(hscode_df)- unique_count_merged : 0
len(mergedHsTariff) - len(tariff_df) : 1129



In [54]:
print((hscode_df['품목번호'].str.len() <= 9).sum()) # 왜 1128개인가?
# '2424000000'가 tariff_df에 누락되어있음. 아마 24년 새로 적용된 HSCODE일 것으로 추정됨

1128


In [68]:
print(hscode_df[hscode_df['품목번호']== '2424000000']['품목번호'])
print("")
print(tariff_df[tariff_df['품목번호']== '2424000000']['품목번호']) # 위 판단에 근거가 됨. tariff_df 공공데이터포털 기준 "최종 수정일 2024-01-04"

2355    2424000000
Name: 품목번호, dtype: object

Series([], Name: 품목번호, dtype: object)


In [56]:

len(hscode_df[hscode_df['품목번호'].str.len() <= 9]) # 1개가 어디서 왔을까?

1128

### tariff에 누락된 관측값 찾는 과정 / '2424000000'인 것으로 확인

In [68]:
unique_merged = set(mergedHsTariff['품목번호'].unique())
unique_hscode = set(hscode_df['품목번호'].unique())

non_duplicates = set(unique_merged.symmetric_difference(unique_hscode))
len(non_duplicates)

1129

In [65]:
missingValue =list(non_duplicates.symmetric_difference(unique_hscode))

In [67]:
len(missingValue)

11293

In [69]:
non_duplicates_set = set(non_duplicates)
unique_hscode = set(hscode_df['품목번호'].unique())

difference = list(non_duplicates_set - unique_hscode)

In [71]:
unique_hscode = set(hscode_df[hscode_df['품목번호'].str.len() <= 9]['품목번호'].unique()) # 1128
unique_merged = set(hscode_df['품목번호'] - mergedHsTariff['품목번호'].unique())

only_in_hscode = list(unique_hscode - unique_merged)

In [75]:
unique_hscode = set(hscode_df['품목번호'].unique())
unique_merged = set(mergedHsTariff['품목번호'].unique())

only_in_hscode = list(unique_hscode - unique_merged)

In [79]:
only_in_hscode

1129

In [80]:
hscode_df[hscode_df['품목번호'].str.len() <= 9]

1128

In [81]:
filtered_hscode = set(hscode_df[hscode_df['품목번호'].str.len() <= 9]['품목번호'].unique())
only_in_hscode_set = set(only_in_hscode)

non_overlap = list(filtered_hscode.symmetric_difference(only_in_hscode_set))

In [82]:
non_overlap #찾았다.

['2424000000']

In [117]:
hscode_df[hscode_df['품목번호']=='2424000000']

품목번호    적용시작일자    적용종료일자 한글품목명            영문품목명 수량단위코드 중량단위코드  \
2355  2424000000  20020101  20241231  이사화물  household goods    NaN     KG   

     수출성질코드 수입성질코드  성질통합분류코드  성질통합분류코드명  
2355  11Z01  14Z01  14010101  (간이세율적용분)

In [118]:
mergedHsTariff[mergedHsTariff['품목번호']=='2424000000']

품목번호 한글품목명            영문품목명  관세율 관세율구분  성질통합분류코드명    적용시작일자  \
60798  2424000000  이사화물  household goods  NaN   NaN  (간이세율적용분)  20020101   

         적용종료일자 수량단위코드 중량단위코드 수출성질코드 수입성질코드  성질통합분류코드 단위당세액 적용국가구분 용도세율구분  \
60798  20241231    NaN     KG  11Z01  14Z01  14010101   NaN    NaN    NaN   

      적용개시일 적용만료일  
60798   NaN   NaN

In [87]:
new = pd.read_csv('HS코드(2024)_new.csv',dtype=str, encoding='cp949')
new[new['HS부호']== '2424000000']

HS부호    적용시작일자    적용종료일자 한글품목명            영문품목명 HS부호내용 한국표준무역분류명  \
2355  2424000000  20020101  20241231  이사화물  household goods    NaN       NaN   

     수량단위최대단가 중량단위최대단가 수량단위코드 중량단위코드 수출성질코드 수입성질코드 품목규격명 필수규격명 참고규격명 규격설명  \
2355      NaN      NaN    NaN     KG  11Z01  14Z01   NaN   NaN   NaN  NaN   

     규격사항내용  성질통합분류코드  성질통합분류코드명  
2355    NaN  14010101  (간이세율적용분)

In [111]:
print(len(new[new['HS부호'].str.len() <= 9]))
#filtered_hscode = set(hscode_df[hscode_df['품목번호'].str.len() <= 9]['품목번호'].unique())
print(len(new['HS부호']))

# 추가 받은 HS부호 파일과 동일한 것으로 확인

1128
12422


### 국가별로 당국 FTA구분을 갖는 행의 값이 다름. ->  해당 값들이 협약상 예외항목일 수도 있고, 누락된 값일 수도 있고, 다른 조약의 적용을 받는 항목일 수도 있음.

In [169]:
print(mergedHsTariff['품목번호'].nunique() - len(mergedHsTariff[mergedHsTariff['관세율구분'] == 'FAU1']))
print(mergedHsTariff['품목번호'].nunique() - len(mergedHsTariff[mergedHsTariff['관세율구분'] == 'FCL1']))
print(mergedHsTariff['품목번호'].nunique() - len(mergedHsTariff[mergedHsTariff['관세율구분'] == 'FCA1']))
print(mergedHsTariff['품목번호'].nunique() - len(mergedHsTariff[mergedHsTariff['관세율구분'] == 'FCN1']))
print(mergedHsTariff['품목번호'].nunique() - len(mergedHsTariff[mergedHsTariff['관세율구분'] == 'FUS1']))

mergedHsTariff['품목번호'].nunique()


1140
1642
1149
1145
1144


12422

### 항목별로 갖는 관세율구분의 개수가 다름 -> 이유는 : 협약 내 포함되는 항목이 아닌 경우 value를 갖지 않음

### 실제로 '관세율구분'에서 특정 value(협약구분기호)를 갖지 않는 항목에 대해서 협약 예외 항목인지를 확인

In [143]:
def check_not_fau1_all(group):
    return (group['관세율구분'] != 'FAU1').all()

not_fau1_all = mergedHsTariff.groupby('품목번호').apply(check_not_fau1_all)
not_fau1_all = not_fau1_all[not_fau1_all]

not_fau1_all_items = not_fau1_all.index.tolist()

In [172]:
def check_fau1_any(group):
    return (group['관세율구분'] == 'FAU1').any()

fau1_at_least_once = mergedHsTariff.groupby('품목번호').apply(check_fau1_any)
fau1_at_least_once = fau1_at_least_once[fau1_at_least_once]

fau1_at_least_once_items = set(fau1_at_least_once.index.tolist())

not_fau1_all_items_set = set(not_fau1_all_items)

result = not_fau1_all_items_set - fau1_at_least_once_items
len(result)

1145

### FAU1이 존재 하지 않는 품목은 1145개 이다. 즉, 호주와의 FTA가 적용되지 않는 품목은 1145개 이다.
#### 가능하다면 협약문을 통해 실제로 적용하지 않은 품목의 개수가 1145개가 맞는지도 확인하면 좋겠지만, 이 방향으로는 정보가 제한적인 것으로 보임..

In [176]:
unique_items = set(mergedHsTariff['품목번호'].unique())
fau1_items = set(mergedHsTariff[mergedHsTariff['관세율구분'] == 'FAU1']['품목번호'].unique())

# print(mergedHsTariff['품목번호'].nunique() - len(mergedHsTariff[mergedHsTariff['관세율구분'] == 'FAU1']))
diff_items = unique_items - fau1_items
result_set = set(result)

not_in_result = diff_items - result_set
not_in_result

set()

In [168]:
diff_items

{'48239090',
 '85042190',
 '72104910',
 '210120901',
 '76169990',
 '72279090',
 '85030020',
 '1006100000',
 '43031029',
 '84798920',
 '16055420',
 '21012090',
 '48149010',
 '23099010',
 '32049010',
 '61099090',
 '53029020',
 '27109140',
 '38221930',
 '32082010',
 '87059090',
 '90112010',
 '440839909',
 '121229703',
 '84135090',
 '382499905',
 '121221802',
 '13021911',
 '29309070',
 '71039990',
 '29309090',
 '44032620',
 '84439110',
 '87044290',
 '84212930',
 '3814002',
 '81019910',
 '15149910',
 '84798910',
 '32082020',
 '848640203',
 '87019110',
 '180690309',
 '330210201',
 '84623210',
 '84472020',
 '850440201',
 '73030010',
 '854449201',
 '90292010',
 '84295110',
 '300692203',
 '283699103',
 '03052040',
 '90328990',
 '88052910',
 '38229000',
 '1211203',
 '16041410',
 '722592909',
 '72104990',
 '84518090',
 '40040010',
 '84799050',
 '85043190',
 '28030090',
 '87021030',
 '83014010',
 '130219903',
 '3824997',
 '55096910',
 '84181010',
 '12122120',
 '85042290',
 '19011010',
 '8414809',


In [181]:
#print((mergedHsTariff[mergedHsTariff['관세율구분'] == 'FAU1']['품목번호']).nunique())
#print((mergedHsTariff[mergedHsTariff['관세율구분'] == 'FAU1']['품목번호']).nunique())
fau1_items = mergedHsTariff[mergedHsTariff['관세율구분'] != 'FAU1']['품목번호']
fau1_items[fau1_items.duplicated()].nunique()

11293

### HSK 22년 기준 11293개(12,242개에서 341개 신설, 1290개 삭제)

In [182]:
print(len(result))
result # FAU1이 없는 품목번호들

1145


{'85042190',
 '48239090',
 '72104910',
 '210120901',
 '76169990',
 '72279090',
 '85030020',
 '1006100000',
 '43031029',
 '84798920',
 '16055420',
 '21012090',
 '48149010',
 '32049010',
 '23099010',
 '61099090',
 '27109140',
 '53029020',
 '38221930',
 '32082010',
 '87059090',
 '90112010',
 '440839909',
 '121229703',
 '84135090',
 '13021911',
 '121221802',
 '382499905',
 '29309070',
 '29309090',
 '71039990',
 '44032620',
 '84439110',
 '87044290',
 '84212930',
 '3814002',
 '81019910',
 '15149910',
 '84798910',
 '32082020',
 '848640203',
 '330210201',
 '180690309',
 '87019110',
 '84623210',
 '84472020',
 '73030010',
 '850440201',
 '854449201',
 '90292010',
 '84295110',
 '300692203',
 '283699103',
 '03052040',
 '90328990',
 '88052910',
 '38229000',
 '1211203',
 '16041410',
 '722592909',
 '72104990',
 '84518090',
 '40040010',
 '28030090',
 '84799050',
 '85043190',
 '87021030',
 '130219903',
 '83014010',
 '3824997',
 '12122120',
 '55096910',
 '84181010',
 '85042290',
 '19011010',
 '21050010',

In [185]:
print(len(mergedHsTariff[mergedHsTariff['품목번호']== '0805100000']))
mergedHsTariff[mergedHsTariff['품목번호']== '0805100000']

30


품목번호 한글품목명    영문품목명 관세율 관세율구분 성질통합분류코드명    적용시작일자    적용종료일자  \
28742  0805100000   오렌지  Oranges  50     A     (오렌지)  20020101  20241231   
28743  0805100000   오렌지  Oranges  50  FAS1     (오렌지)  20020101  20241231   
28744  0805100000   오렌지  Oranges  50  FAU1     (오렌지)  20020101  20241231   
28745  0805100000   오렌지  Oranges  50  FAU1     (오렌지)  20020101  20241231   
28746  0805100000   오렌지  Oranges   0  FAU2     (오렌지)  20020101  20241231   
28747  0805100000   오렌지  Oranges   0  FAU8     (오렌지)  20020101  20241231   
28748  0805100000   오렌지  Oranges  50  FCA1     (오렌지)  20020101  20241231   
28749  0805100000   오렌지  Oranges  50  FCN1     (오렌지)  20020101  20241231   
28750  0805100000   오렌지  Oranges  50  FCO1     (오렌지)  20020101  20241231   
28751  0805100000   오렌지  Oranges  50  FEU1     (오렌지)  20020101  20241231   
28752  0805100000   오렌지  Oranges   0  FEU2     (오렌지)  20020101  20241231   
28753  0805100000   오렌지  Oranges   0  FEU6     (오렌지)  20020101  20241231   
28754  0805100000   오렌지  Oranges  50  FGB1     (오렌지)  20020101  20241231   
28755  0805100000   오렌지  Oranges   0  FGB2     (오렌지)  20020101  20241231   
28756  0805100000   오렌지  Oranges  50  FIL1     (오렌지)  20020101  20241231   
28757  0805100000   오렌지  Oranges  50  FKH1     (오렌지)  20020101  20241231   
28758  0805100000   오렌지  Oranges  50  FNZ1     (오렌지)  20020101  20241231   
28759  0805100000   오렌지  Oranges  50  FPE1     (오렌지)  20020101  20241231   
28760  0805100000   오렌지  Oranges  50  FPE1     (오렌지)  20020101  20241231   
28761  0805100000   오렌지  Oranges   0  FPE2     (오렌지)  20020101  20241231   
28762  0805100000   오렌지  Oranges  50  FTR1     (오렌지)  20020101  20241231   
28763  0805100000   오렌지  Oranges  50  FUS1     (오렌지)  20020101  20241231   
28764  0805100000   오렌지  Oranges  50  FUS1     (오렌지)  20020101  20241231   
28765  0805100000   오렌지  Oranges   0  FUS2     (오렌지)  20020101  20241231   
28766  0805100000   오렌지  Oranges   0  FUS6     (오렌지)  20020101  20241231   
28767  0805100000   오렌지  Oranges   0  FUS6     (오렌지)  20020101  20241231   
28768  0805100000   오렌지  Oranges  50  FVN1     (오렌지)  20020101  20241231   
28769  0805100000   오렌지  Oranges   0     U     (오렌지)  20020101  20241231   
28770  0805100000   오렌지  Oranges  50    W1     (오렌지)  20020101  20241231   
28771  0805100000   오렌지  Oranges  50    W2     (오렌지)  20020101  20241231   

      수량단위코드 중량단위코드 수출성질코드 수입성질코드  성질통합분류코드 단위당세액 적용국가구분 용도세율구분       적용개시일  \
28742    NaN     KG  11Z01  12399  11010603   NaN      1    NaN  2024-01-01   
28743    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-01-01   
28744    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-01-01   
28745    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-10-01   
28746    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-04-01   
28747    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-01-01   
28748    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-01-01   
28749    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-01-01   
28750    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-01-01   
28751    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-01-01   
28752    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-03-01   
28753    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-01-01   
28754    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-01-01   
28755    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-03-01   
28756    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-01-01   
28757    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-01-01   
28758    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-01-01   
28759    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-01-01   
28760    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-11-01   
28761    NaN     KG  11Z01  12399  11010603   NaN      2    NaN  2024-05-01 

### 세율 적용 우선 순위에 해당하는 아래항목 중 L(조정관세)만 관세율구분에 존재함

In [187]:
items = ['I', 'M', 'J', 'K', 'T', 'L']
existing_items = [item for item in items if item in mergedHsTariff['관세율구분'].values]
existing_items

['L']

In [190]:
L_items = mergedHsTariff[mergedHsTariff['관세율구분'] == 'L']['품목번호']
L_items.unique()


array(['0301929090', '0301994091', '0301994099', '0301999059',
       '0303592000', '0303670000', '0303699000', '0306951030',
       '0306959030', '0307432010', '0307432090', '0709540000',
       '0712340000', '1902192000', '1904901010', '2103901030',
       '2103901090', '2103909020', '2103909030', '2103909090',
       '4412314019', '4412314021', '4412314029', '4412315010',
       '4412315090', '4412316010', '4412316090', '4412317010',
       '4412317090', '4412334010', '4412334020', '4412335000',
       '4412336000', '4412337000', '4412344010', '4412344020',
       '4412345000', '4412346000', '4412347000', '4412399000',
       '4412411000', '4412421000', '4412491000', '4412911000',
       '4412919110', '4412919910', '4412921000', '4412929110',
       '4412929910', '4412991000', '4412999910'], dtype=object)

In [ ]:
def check_fau1_any(group):
    return (group['관세율구분'] == 'FAU1').any()

fau1_at_least_once = mergedHsTariff.groupby('품목번호').apply(check_fau1_any)
fau1_at_least_once = fau1_at_least_once[fau1_at_least_once]

fau1_at_least_once_items = set(fau1_at_least_once.index.tolist())

not_fau1_all_items_set = set(not_fau1_all_items)

result = not_fau1_all_items_set - fau1_at_least_once_items
len(result)

1145

In [ ]:
def check_fau1_any(group):
    return (group['관세율구분'] == 'FAU1').any()

fau1_at_least_once = mergedHsTariff.groupby('품목번호').apply(check_fau1_any)
fau1_at_least_once = fau1_at_least_once[fau1_at_least_once]

fau1_at_least_once_items = set(fau1_at_least_once.index.tolist())

not_fau1_all_items_set = set(not_fau1_all_items)

result = not_fau1_all_items_set - fau1_at_least_once_items
len(result)

1145

In [ ]:
def check_fau1_any(group):
    return (group['관세율구분'] == 'FAU1').any()

fau1_at_least_once = mergedHsTariff.groupby('품목번호').apply(check_fau1_any)
fau1_at_least_once = fau1_at_least_once[fau1_at_least_once]

fau1_at_least_once_items = set(fau1_at_least_once.index.tolist())

not_fau1_all_items_set = set(not_fau1_all_items)

result = not_fau1_all_items_set - fau1_at_least_once_items
len(result)

1145

In [205]:
data_df = hscode_df[['품목번호','적용시작일자','적용종료일자']]

In [206]:
data_df['적용시작일자'] = pd.to_datetime(data_df['적용시작일자'])
data_df['적용종료일자'] = pd.to_datetime(data_df['적용종료일자'])

earliest_start_date = data_df['적용시작일자'].min()
latest_end_date = data_df['적용종료일자'].max()

/var/folders/4j/3hzvjkrn3t722p4ygj5c7zq80000gn/T/ipykernel_52164/583311680.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_df['적용시작일자'] = pd.to_datetime(data_df['적용시작일자'])
/var/folders/4j/3hzvjkrn3t722p4ygj5c7zq80000gn/T/ipykernel_52164/583311680.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_df['적용종료일자'] = pd.to_datetime(data_df['적용종료일자'])


In [207]:
print(data_df['적용시작일자'].min())
print(data_df['적용종료일자'].max())

data_df[data_df['적용시작일자'] == '2002-01-01 00:00:00']

2002-01-01 00:00:00
2024-12-31 00:00:00


품목번호     적용시작일자     적용종료일자
22     0103910000 2002-01-01 2024-12-31
23     0103920000 2002-01-01 2024-12-31
27     0104209000 2002-01-01 2024-12-31
29     0105119000 2002-01-01 2024-12-31
30     0105120000 2002-01-01 2024-12-31
...           ...        ...        ...
12370  9616200000 2002-01-01 2024-12-31
12372  9617002000 2002-01-01 2024-12-31
12373  9617008000 2002-01-01 2024-12-31
12374  9617009000 2002-01-01 2024-12-31
12376  9618009000 2002-01-01 2024-12-31

[3810 rows x 3 columns]